Start blocking the ability to add local bare repositories

Daniel-McCarthy opened the pull request:
This pull request is in response to Issue #4293. The intention is to stop allowing bare repositories from being added via GitHub Desktop as they are not currently supported. To do so, a function that checks if the repository is bare has been added. Additionally, the Add-Existing-Repository dialogue now uses the added function to check for a bare repository and displays a warning and disables the add button.

I am still new to this project and a number of the technologies involved in it so there is likely plenty room for improvement. I am more than happy to make changes to my implementation. I suspect it can be simplified, especially where it deals with `boolean | null` types.

shiftkey commented on app/src/lib/git/rev-parse.ts:
```diff
@@ -52,6 +52,32 @@ export async function getTopLevelWorkingDirectory(
   return Path.resolve(path, relativePath)
 }
 
+export async function checkIfRepositoryIsBare(
```
:art: can we rename this function to `isBareRepository` - this feels more On Brand™ alongside `isGitRepository` which is also defined in this file.

shiftkey commented on app/src/lib/git/rev-parse.ts:
```diff
@@ -52,6 +52,32 @@ export async function getTopLevelWorkingDirectory(
+    result = await git(
+      ['rev-parse', '--is-bare-repository'],
+      path,
+      'checkIfRepositoryIsBare'
```
:art: this should match the function being called

shiftkey replied:
This `'checkIfRepositoryIsBare'` string is used for tracing Git operations (in case they run unexpectedly long or if they error, details end up in the log). The convention is that this name should be the same as the function itself.

Daniel-McCarthy replied:
The source of my confusion is that the function I have added is named checkIfRepositoryIsBare and I am passing a string 'checkIfRepositoryIsBare'.

Similar to how getTopLevelWorkingDirectory passes 'getTopLevelWorkingDirectory' as an argument.

Did I make a mistake with the string? One thing I noticed that seems relevant is that there is 'isGitRepostory' and 'getTopLevelWorkingDirectory', with the prior calling the latter. I did not copy this pattern and only have checkIfRepostoryIsBare. Perhaps that is part of my problem.

Sorry to take up your time with this.

shiftkey replied:
We can't follow this pattern for `isGitRepository` because it defers to `getTopLevelWorkingDirectory` for it's Git operation.

The pattern we're trying to follow here is **if a function performs a Git operation, ensure it's named in a way that we can trace back to later**.

What I was looking to ensure here is that the function name itself is in sync with the string we pass into `git()`, and I think we're on the same page. I just wanted to ensure the function name change mentioned above https://github.com/desktop/desktop/pull/5434#discussion_r213014336 is also applied here.

shiftkey commented on app/src/lib/git/rev-parse.ts:
```diff
@@ -52,6 +52,32 @@ export async function getTopLevelWorkingDirectory(
+    )
+  } catch (e) {
+    if (e.message.includes('not a git repository')) {
+      return null
```
Are you able to add some documentation to the header of this function, so we can understand when we should expect `null` versus a boolean versus an error?

shiftkey commented on app/src/ui/add-repository/add-existing-repository.tsx:
```diff
@@ -158,11 +183,18 @@ export class AddExistingRepository extends React.Component<
 
     const path = directory[0]
     const isRepository = await isGitRepository(path)
+    const isBareRepositoryResult = await checkIfRepositoryIsBare(path)
```
:art: I think we can distill this down into a one-liner, based on what I think `null` represents:

```ts
const isBareRepository = (await checkIfRepositoryIsBare(path)) || false
```

shiftkey commented on app/src/lib/git/rev-parse.ts:
```diff
@@ -52,6 +52,32 @@ export async function getTopLevelWorkingDirectory(
+  if (result === null) {
+    return null
+  } else {
+    return result.stdout === 'true\n'
```
:art: use `.trim()` to avoid the need to check for `\n` here

shiftkey commented on app/src/lib/git/rev-parse.ts:
```diff
@@ -52,6 +52,32 @@ export async function getTopLevelWorkingDirectory(
+    throw e
+  }
+
+  if (result === null) {
```
Rather than trying to make the try-catch as narrow as possible, let's just lift this up into the try-catch to simplify things:

```
  try {
    const result = await git(
      ['rev-parse', '--is-bare-repository'],
      path,
      'checkIfRepositoryIsBare'
    )
    return result.stdout.trim() === 'true'
  } catch (e) {
    if (e.message.includes('not a git repository')) {
      return null
    }

    throw e
  }
```

Daniel-McCarthy commented:
Okay I have updated this pull requests changes in response to the review. Thanks again @shiftkey for the feedback and assistance!

I believe I have addressed each of the requested changes:
-The ```checkIfRepositoryIsBare``` function in the /lib/git/rev-parse.ts is now named ```isBareRepository```.
-The string being passed in ```isBareRepository``` has been updated to match the name change.
-I added documentation above the ```isBareRepository``` function addressing the null return.
-I made use of the trim() function instead of including ```\n``` in the string.
-And I have made the simplification adding the ```|| false```.

Additionally I updated the unit tests to make use of the renamed function's new name.

iAmWillShepherd commented on app/src/lib/git/rev-parse.ts:
```diff
@@ -52,6 +52,30 @@ export async function getTopLevelWorkingDirectory(
+    return result.stdout.trim() === 'true'
+  } catch (e) {
+    if (e.message.includes('not a git repository')) {
+      return null
```
Curious why not just return false here? Given the function name, I'd expect a yes/no answer and not really care too much about the reason why it's not a bare repo.

Daniel-McCarthy replied:
That's a great point. Thanks for the feedback!

I would be happy to update the pull request again. To clarify, I should make this function return a ```Promise<boolean>``` instead of a ```Promise<boolean | null>``` and remove the "@returns null ..." section I added to the documentation, and cleanup the section mentioned in the add-existing-repository.tsx?

I just want to be on the same page regarding the full extent of the suggested changes.

Daniel-McCarthy replied:
@shiftkey 
I am looking at implementing @iAmWillShepherd 's recommendations. However, since the method would no longer return null it would also cause one of the unit tests you have added to fail.

Should I go ahead with this and remove that particular unit test? If so then I shall do so, double check my changes, and then update the pull request.

shiftkey replied:
If the test no longer catches a case we're interested in, removing it is fine

Daniel-McCarthy replied:
I see. I think it might make more sense to change the unit test to check if it returns false since it is still a potential result from the function. I can update the test with that change and verify the tests still function as expected.

iAmWillShepherd commented on app/src/ui/add-repository/add-existing-repository.tsx:
```diff
@@ -82,14 +85,33 @@ export class AddExistingRepository extends React.Component<
     }
 
+    const isBare = await isBareRepository(this.state.path)
+    if (isBare !== null && isBare === true) {
```
Returning just a `bool` cleans this up as `null` doesn't seem to serve any purpose here.

shiftkey commented on app/src/lib/git/rev-parse.ts:
```diff
@@ -52,6 +52,30 @@ export async function getTopLevelWorkingDirectory(
+ *
+ * @param path The path to the Git repository to check.
+ *
+ * @returns null if the path provided does not contain a Git repository.
```
This comment needs to be updated to indicate what `true` and `false` represent

Daniel-McCarthy replied:
Ah, of course. I had forgotten to make that change before I made the commit. I'll fix that now.

shiftkey commented on app/src/lib/git/rev-parse.ts:
```diff
@@ -52,6 +52,30 @@ export async function getTopLevelWorkingDirectory(
 }
 
+/**
+ * Checks if the repository at a path is bare.
```
It looks like this code can handle a non-Git repository being at the path, but it errors if the path doesn't exist on disk. Can we capture that here and in the `@param path` value below?

Daniel-McCarthy replied:
@shiftkey 
In regards to these two changes requested, would the following be sufficient wording? If so I will submit it in another commit.

```
/**
 * Checks if the repository at a path is bare.
 *
 * @param path The path to the repository to check. An error will be thrown if the path does not exist on disk.
 *
 * @returns true if the path contains a bare Git repository. Returns false if it is not bare or is not a Git repository.
 */
```

shiftkey replied:
@Daniel-McCarthy works for me

searchgold commented:
5207
